File: apps/frontend/src/types/overview.ts
export interface OverviewRepositoryItem {
  id: number;
  name: string;
  full_name: string;
  description: string | null;
  default_branch: string;
  primary_language: string | null;
  latest_analysis_status: string | null;
  latest_analysis_id: number | null;
  loc: number;
  files: number;
  health_score: number;
}

export interface OverviewIssueItem {
  id: string | number;
  repo_name: string;
  repo_id: number;
  title: string;
  description: string;
  severity: string;
  type: string;
  status?: string;
  dismissed_reason?: string | null;
  resolved_at?: string | null;
}

export interface OverviewRecommendationItem {
  id: string | number;
  repo_name: string;
  repo_id: number;
  recommendation: string;
  priority: string;
  status?: string;
  action_plan?: string | null;
  resolved_at?: string | null;
  component_id?: string | null;
  component_name?: string | null;
  subsystem?: string | null;
  category?: string | null;
}

export interface OverviewRecentActivity {
  timestamp: string;
  text: string;
  tag: "good" | "warn" | "risk";
}

export interface OverviewTechnology {
  name: string;
  count: number;
  confidence: number;
}

export interface HealthBreakdown {
  maintainability: number;
  complexity: number;
  coupling: number;
  modularity: number;
}

export interface OrganizationOverviewResponse {
  organization_id: number;
  total_repos: number;
  analyzed_repos_count: number;
  total_loc: number;
  total_files: number;
  total_classes: number;
  total_functions: number;
  health_score: number;
  health_label: "Healthy" | "Watch" | "At risk";
  risk_level: "Low" | "Medium" | "High";
  critical_findings_count: number;
  warning_findings_count: number;
  last_snapshot_iso: string | null;
  health_breakdown: HealthBreakdown;
  repos: OverviewRepositoryItem[];
  all_issues: OverviewIssueItem[];
  all_recommendations: OverviewRecommendationItem[];
  recent_activities: OverviewRecentActivity[];
  technologies: OverviewTechnology[];
}

File: apps/backend/app/schemas/overview.py
"""
Organization Overview Telemetry Schemas.

Provides high-speed, aggregated telemetry models for the Coodara
dashboard, risks, recommendations, and reports pages.
"""

from __future__ import annotations

from typing import Literal
from pydantic import BaseModel, ConfigDict, Field


class OverviewRepositoryItem(BaseModel):
    """Repository snapshot metadata within an organization."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    full_name: str
    description: str | None = None
    default_branch: str
    primary_language: str | None = None
    latest_analysis_status: str | None = None
    latest_analysis_id: int | None = None
    loc: int = 0
    files: int = 0
    health_score: float = 100.0


class OverviewIssueItem(BaseModel):
    """Architectural issue summary item."""

    model_config = ConfigDict(from_attributes=True)

    id: str | int
    repo_name: str
    repo_id: int
    title: str
    description: str
    severity: str
    type: str
    status: str = "open"
    dismissed_reason: str | None = None
    resolved_at: str | None = None


class OverviewRecommendationItem(BaseModel):
    """Architectural recommendation summary item."""

    model_config = ConfigDict(from_attributes=True)

    id: str | int
    repo_name: str
    repo_id: int
    recommendation: str
    priority: str
    component_id: str | None = None
    component_name: str | None = None
    subsystem: str | None = None
    category: str | None = None
    status: str = "open"
    action_plan: str | None = None
    resolved_at: str | None = None


class OverviewRecentActivity(BaseModel):
    """Timeline event item."""

    timestamp: str
    text: str
    tag: Literal["good", "warn", "risk"]


class OverviewTechnology(BaseModel):
    """Aggregated technology item."""

    name: str
    count: int
    confidence: float


class HealthBreakdown(BaseModel):
    """Structural health sub-metrics."""

    maintainability: float = 85.0
    complexity: float = 80.0
    coupling: float = 75.0
    modularity: float = 85.0


class OrganizationOverviewResponse(BaseModel):
    """
    Consolidated, pre-aggregated telemetry for an organization.
    Powers Overview, Risks, Recommendations, Reports, and Sidebar in a single query.
    """

    organization_id: int
    total_repos: int = 0
    analyzed_repos_count: int = 0
    total_loc: int = 0
    total_files: int = 0
    total_classes: int = 0
    total_functions: int = 0
    health_score: float = 100.0
    health_label: Literal["Healthy", "Watch", "At risk"] = "Healthy"
    risk_level: Literal["Low", "Medium", "High"] = "Low"
    critical_findings_count: int = 0
    warning_findings_count: int = 0
    last_snapshot_iso: str | None = None
    health_breakdown: HealthBreakdown = Field(default_factory=HealthBreakdown)
    repos: list[OverviewRepositoryItem] = Field(default_factory=list)
    all_issues: list[OverviewIssueItem] = Field(default_factory=list)
    all_recommendations: list[OverviewRecommendationItem] = Field(default_factory=list)
    recent_activities: list[OverviewRecentActivity] = Field(default_factory=list)
    technologies: list[OverviewTechnology] = Field(default_factory=list)
